rest_api/aggregation/database_hooks.py

```diff
@@ -0,0 +1,345 @@
+
+    @staticmethod
+    def _get_relation(endpoint, where):
+        d = db[endpoint].find_one(where)
+        if d:
+            return data_relations[endpoint](d)
+
+    def _embed(self):
+        for k, v in self.subelements.items():
+            rels = self._get_relations(k, v)
+            self.data[k] = [r.data for r in rels]
+
+    def _trigger_superelement_aggregation(self):
+        for k, v in self.superelements.items():
+            rel = self._get_relation(k, v)
+            if rel:
+                rel.aggregate()
+
+    def aggregate(self):
+        self._embed()
+        if self.aggregated_fields:
+            self.data['aggregated'] = {}
+            for stage in self.aggregated_fields:
+                new_data = resolve(stage, self.data)
+                self.data['aggregated'].update(new_data)
+
+            db[self.endpoint].update_one(
+                {self.id_field: self.data[self.id_field]},
+                {'$set': {'aggregated': self.data['aggregated']}}
+            )
+
+        self._trigger_superelement_aggregation()
+
+
+# TODO: move these alongside the rest of the expressions in aggregation.database_side
+class Reference(Calculation):
+    def _expression(self, field):
+        return field
+
+
+class Mean(Accumulation):
+    def _expression(self, elements):
+        if elements:
+            return sum(elements) / len(elements)
+
+
+class FirstElement(Accumulation):
+    def _expression(self, elements):
+        if elements:
+            return elements[0]
+
+
+class StDevPop(Accumulation):
+    def _expression(self, elements):
+        if elements:
+            return statistics.pstdev(elements)
+
+
+class GenotypeMatch(Calculation):
+    def _expression(self, geno_val):
+        if not geno_val:
+            return None
+        elif geno_val['no_call_chip'] + geno_val['no_call_seq'] < 15:
+            if geno_val['mismatching_snps'] < 6:
+                return 'Match'
+            elif geno_val['mismatching_snps'] > 5:
+                return 'Mismatch'
+        else:
+            return 'Unknown'
+
+
+class SexCheck(Calculation):
+    def _expression(self, called, provided):
+        if not called or not provided:
+            return None
+        elif called == provided:
+            return called
+        else:
+            return 'Mismatch'
+
+
+class MatchingSpecies(Calculation):
+    def _expression(self, species_contam):
+        return sorted(k for k, v in species_contam['contaminant_unique_mapped'].items() if v > 500)
+
+
+class MostRecent(Calculation):  # TODO: Should replace server_side.MostRecent
+    def __init__(self, *args, date_field='_created'):
+        self.date_field = date_field
+        super().__init__(*args)
+
+    def _expression(self, elements):
+        procs = sorted(elements, key=lambda x: x.get(self.date_field))
+        if procs:
+            return procs[-1]
+
+
+class NbUniqueMutableElements(Calculation):  # TODO: Should replace server_side.NbUniqueElements
```

Comment:
Don't understand why "Mutable". Did you mean "Iterable"?

Reply:
Yeah, I just wanted something named differently from server_side.expressions.NbUniqueElements to avoid them conflicting. The behaviour needs to be different here because it was doing `set` on a list of dicts and getting a TypeError: unhashable type. Can rename?